crates/valence_nbt/src/snbt.rs

```diff
@@ -0,0 +1,685 @@
+        }
+    }
+
+    fn parse_primitive(&mut self) -> Result<Value> {
+        macro_rules! try_ret {
+            // Try possible solution until one works
+            ($v:expr) => {{
+                match $v {
+                    Ok(v) => return Ok(v.into()),
+                    Err(_) => (),
+                }
+            }};
+        }
+        let target = self.read_unquoted_string()?;
+        match target
+            .bytes()
+            .last()
+            .ok_or_else(|| self.make_error(SNBTErrorType::ExpectValue))?
+        {
+            b'b' | b'B' => try_ret!(target[..target.len() - 1].parse::<i8>()),
+            b's' | b'S' => try_ret!(target[..target.len() - 1].parse::<i16>()),
+            b'l' | b'L' => try_ret!(target[..target.len() - 1].parse::<i64>()),
+            b'f' | b'F' => try_ret!(target[..target.len() - 1].parse::<f32>()),
+            b'd' | b'D' => try_ret!(target[..target.len() - 1].parse::<f64>()),
+            _ => (),
+        }
+        match target.as_str() {
+            "true" => return Ok(Value::Byte(1)),
+            "false" => return Ok(Value::Byte(0)),
+            _ => {
+                try_ret!(target.parse::<i32>());
+                try_ret!(target.parse::<f64>());
+            }
+        };
+        if target.len() > STRING_MAX_LEN {
+            return Err(self.make_error(SNBTErrorType::LongString));
+        }
+        Ok(Value::String(target))
+    }
+
+    /// Read the next element in the SNBT string.
+    /// [`SNBTErrorType::TrailingData`] is impossible to be returned since it doesnot consider it's an error.
+    pub fn parse_element(&mut self) -> Result<Value> {
+        self.skip_whitespace();
+        match self.peek()? {
+            '{' => self.parse_compound().map(|c| c.into()),
+            '[' => self.parse_list_like(),
+            '"' | '\'' => self.read_quoted_string().map(|s| s.into()),
+            _ => self.parse_primitive(),
+        }
+    }
+
+    pub fn read(&mut self) -> Result<Value> {
+        let value = self.parse_element()?;
+        self.skip_whitespace();
+        if self.peek().is_ok() {
+            return Err(self.make_error(SNBTErrorType::TrailingData));
+        }
+        Ok(value)
+    }
+
+    /// Get the number of bytes readed.
+    /// It's useful when you want to read a SNBT string from an command argument since there may be trailing data.
+    pub fn bytes_readed(&self) -> usize {
+        self.index
+    }
+
+    /// Parse a string in SNBT format into a `Value`.
+    /// Assert that the string has no trailing data.
+    /// SNBT is quite similar to JSON, but with some differences.
+    /// See [the wiki](https://minecraft.gamepedia.com/NBT_format#SNBT_format) for more information.
+    /// # Example
+    /// ```
+    /// use nbt::Value;
+    /// use nbt::SNBTReader;
+    /// let value = SNBTReader::from_snbt("1f").unwrap();
+    /// assert_eq!(value, Value::Float(1.0));
+    /// ```
+    pub fn from_snbt(snbt: &str) -> Result<Value> {
```

Comment:
Rather than have this as a function on `SnbtReader`, it might be best to export a free function like `from_snbt_str` at the root of the crate. This would be consistent with `from_binary_slice` and `to_binary_writer`. Same goes for `SnbtWriter::stringify`, except that it should be named `to_snbt_string`.